### ISSUE TITLE
Error when collecting count of agents where a model initialises with none of a certain type.

### ISSUE THREAD

**Issue by jacobusmmsmit** [JuliaDynamics/Agents.jl#446]
**Describe the bug**
Initialising a model with 0 agent of a certain type, then passing `adata` to the `run!` function where you count the number of each type gives this error:
`ERROR: ArgumentError: collection must be non-empty`

**Minimal Working Example**
In the `predator-prey` example, you can set `n_wolves = 0` and the error will appear when run.
```
model = initialize_model(
    n_wolves = 0,
    dims = (25, 25),
    Δenergy_sheep = 5,
    sheep_reproduce = 0.2,
    wolf_reproduce = 0.08,
)
results, _ = run!(model, agent_step!, n_steps; adata)
```

**Agents.jl version**

Agents v4.1.3, this was _not_ an issue in Agents v3.7 as this is what I was previously using before upgrading and having this break.

**Comment by Libbum** [comment]
What is `adata` for this case? Could be a bug, but perhaps this was a conscious change we made for 4.x, it'll depend on your `adata`.

**Comment by jacobusmmsmit** [comment]
`adata` is defined as in the example code:
```
sheep(a) = typeof(a) == Sheep
wolves(a) = typeof(a) == Wolf
grass(a) = typeof(a) == Grass && a.fully_grown
adata = [(sheep, count), (wolves, count), (grass, count)]
```

**Comment by Libbum** [comment]
OK, thanks&mdash;I'll take a look.

**Comment by Libbum** [comment]
I quickly identified the error just before bed last night. Looking at it a little more closely this morning, this may be a more nuanced problem than originally anticipated.

The behavior we should expect here is something more like the single case:
```julia
mutable struct Agent3 <: AbstractAgent
    id::Int
    pos::Tuple{Int,Int}
    weight::Float64
end

model = ABM(Agent3, GridSpace((5,5)))
agent3(a) = a isa Agent3
adata = [(agent3, count)]

julia> data, _ = run!(model, dummystep, 1; adata)
ERROR: ArgumentError: Model must have at least one agent to initialize data collection
```

So the current argument error is not helpful and needs to be fixed at the very least. If the model has no instance of an agent type it has no way of identifying the properties of that agent in the general case.

In v3.7, multiagent models like predator prey were not properly supported by the collection routines, whereas they are now. This allowed for an edge case that you're using in 3.7, where the model has one or more agent, but not one or more agent of each type.

I'd need to know a little more about your motivations in this example to be able to provide functionality for it. This line of questioning goes for single-agent models too for the most part.
- Will there ever be wolves in your model? 
   If not, it seems like a better alternative is to remove them from the model entirely.
- Will there be wolves in your model, just not at the start?
  There are a few ways to do this, depending on how things need to be set up.
  1. Initialize the model with at least one agent of each type, call with `run!`, but `kill_agents!` on the first tick of the model (not ideal)
  2. Use the [Data collection](https://juliadynamics.github.io/Agents.jl/stable/api/#Data-collection) interface instead of `run!`.

**Comment by Datseris** [comment]
I don't agree that this is a bug, or something to be fixed. As Julia says, `collection must be non-empty`, You need at least one agent to collect agent data, that's normal to expect.

We can change the error to say `Model must have at least one agent to initialize data collection` instead if `adata` is not `nothing` and `nagents(model) == 0`, but that's about it.

**Comment by Datseris** [comment]
Tim's last idea (using the data collection interface) for a specific case where a model *must* start without any agents is the best solution. The convenience of `run!` cannot be applied in this case.

**Comment by jacobusmmsmit** [comment]
Thank you for the responses. In my case there will be agents of all type in my simulation, just perhaps not at the beginning. I am trying to construct a quiver plot of the one step population makeup over a triangle simplex of possible population makeups (I have three types in my model, hence triangle simplex).

![image](https://user-images.githubusercontent.com/57063134/109988886-dc551400-7cff-11eb-99fd-2a327b8a15e9.png)

From the image you can see that this needs to start sometimes where there are indeed no agents of a certain type. I was thinking of doing this with a `paramscan` but I'm not sure how best to do so.

**Comment by jacobusmmsmit** [comment]
For reference here is how I am currently doing this in Agents.jl v3.7
```
sitters(a) = typeof(a) == Sitter
identifiers(a) = typeof(a) == Identifier
cheaters(a) = typeof(a) == Cheater
adata = [(sitters, count), (identifiers, count), (cheaters, count)]

total_agents = 30
for i in 0:total_agents, j in 0:total_total_agents - i
    initialise_agents!(model, i, j, total_agents - i - j)

    results, _ = run!(model, agent_step!, model_step!, 1; adata=adata, replicates=100);
    
    results = results |>
        df -> filter(:step => !=(0), df) |>
        df -> combine(df, :count_sitters => mean => :xend, :count_identifiers => mean => :yend)
    results[:, :xstart] .= i
    results[:, :ystart] .= j
    final_results = [final_results; results]
end
```
It is rather rudimentary.

**Comment by Libbum** [comment]
Since we cannot know how your (or anyone else's) agents look before hand, we cannot provide this service in general via `run!`, so you'll need to swap that out with a custom run function and use the data collection interface.

Using the [current implementation](https://github.com/JuliaDynamics/Agents.jl/blob/5947d0f3b5defced4894a72b7f0d61017875a2c8/src/simulations/collect.jl#L125) as a basis, something like:

```julia
model = init_function(...) # Add at least one agent of each type
results = init_agent_dataframe(model, adata) # This gets us the initial table layout correct
total_agents = 30
for i in 0:total_agents, j in 0:total_total_agents - i
    # My understanding is that this function calls `genocide!`, so initial agents will be cleared
    initialise_agents!(model, i, j, total_agents - i - j)

    # `results` will be updated
    custom_run!(results, model, agent_step!, model_step!, 1; adata)
    ...
end

function custom_run!(df_agent, model, agent_step!, model_step!, n; adata=nothing)
    s = 0
    while s < n
        collect_agent_data!(df_agent, model, adata, s)
        step!(model, agent_step!, model_step!, 1)
        s += 1
    end
    collect_agent_data!(df_agent, model, adata, s)
    return nothing
end
```

[Replicates](https://github.com/JuliaDynamics/Agents.jl/blob/5947d0f3b5defced4894a72b7f0d61017875a2c8/src/simulations/collect.jl#L434) are handled this way. How this is extended nicely for cases like this is in the plans for the future. We've been working recently on seeding #420 to get replicates to work better #415. For the moment though, a custom implementation that wraps your `custom_run!` is the best bet to get this working in v4.x.

**Comment by Libbum** [comment]
@jacobusmmsmit I'm coming back to this issue soon since we just sorted out some big questions about seeding and replicates. I think we can now do what you're asking a bit easier.

**Comment by mingR** [comment]
> [this line](https://github.com/JuliaDynamics/Agents.jl/blob/1efb62bdc135ea263b2b3aefbf71fc42c1b82ca9/src/simulations/collect.jl#L222) is the problem. It assumes at least one agent of each type in the model. Should be a quick fix.

The error message has been confusing, finally found the answer here, thanks!
